Jellyfin.Server.Implementations/MediaSegments/MediaSegmentManager.cs

```diff
@@ -139,18 +139,47 @@ public async Task DeleteSegmentAsync(Guid segmentId)
     }
 
     /// <inheritdoc />
-    public async Task<IEnumerable<MediaSegmentDto>> GetSegmentsAsync(Guid itemId, IEnumerable<MediaSegmentType>? typeFilter)
+    public Task<IEnumerable<MediaSegmentDto>> GetSegmentsAsync(Guid itemId, IEnumerable<MediaSegmentType>? typeFilter, bool filterByProvider = true)
+    {
+        var baseItem = _libraryManager.GetItemById(itemId);
+
+        if (baseItem is null)
+        {
+            _logger.LogError("Tried to request segments for an invalid item");
+            return Task.FromResult<IEnumerable<MediaSegmentDto>>([]);
+        }
+
+        return GetSegmentsAsync(baseItem, typeFilter, filterByProvider);
+    }
+
+    /// <inheritdoc />
+    public async Task<IEnumerable<MediaSegmentDto>> GetSegmentsAsync(BaseItem item, IEnumerable<MediaSegmentType>? typeFilter, bool filterByProvider = true)
     {
         using var db = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
 
         var query = db.MediaSegments
-            .Where(e => e.ItemId.Equals(itemId));
+            .Where(e => e.ItemId.Equals(item.Id));
 
         if (typeFilter is not null)
         {
             query = query.Where(e => typeFilter.Contains(e.Type));
         }
 
+        if (filterByProvider)
+        {
+            var libraryOptions = _libraryManager.GetLibraryOptions(item);
+            var providers = _segmentProviders
+                .Where(e => !libraryOptions.DisabledMediaSegmentProviders.Contains(GetProviderId(e.Name)))
+                .ToArray();
+            if (providers.Length == 0)
+            {
+                return [];
+            }
+
+            var providerIds = providers.Select(f => GetProviderId(f.Name)).ToArray();
+            query = query.Where(e => providerIds.Contains(e.SegmentProviderId));
+        }
+
         return query
             .OrderBy(e => e.StartTicks)
             .AsNoTracking()
```

Comment:
Unrelated to the PR: Why is there a `ToImmutableList()` followed by a `Select` below this line? The immutability isn't necessary here and it is basically discarded by the `Select`

Reply:
Fixed by using AsEnumerable to enumerate the EfCore query.